### ISSUE TITLE
GPU memory usage does not increase when using multithreading in python 

### ISSUE THREAD

**Issue by Neronjust2017** [tensorflow/tensorflow#50611]
Hi, I am working on deploy some yolo v3 models (keras) on the GPU of our server. However, I found that, when I use multithreading programming in python to deploy multiple models and let them predict in parallel, the GPU memory usage does not increase, the GPU memory usage is almost as much as one single model, which makes me very confused. Could someone help me?  

load and predict one by one: 
**model 0
gpu usage：total 16280.88 MB， used 859.56 MB， unused 15421.31 MB
model 1
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB
model 2
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB
model 3
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB

2021-07-05 16:45:00.227119: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7

gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
**

load one by one and predict through multithread: 
**model 0
gpu usage：total 16280.88 MB， used 861.56 MB， unused 15419.31 MB
model 1
gpu usage：total 16280.88 MB， used 1373.56 MB， unused 14907.31 MB
model 2
gpu usage：total 16280.88 MB， used 1373.56 MB， unused 14907.31 MB
model 3
gpu usage：total 16280.88 MB， used 2397.56 MB， unused 13883.31 MB

2021-07-05 17:06:22.285915: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7

thread 0 gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 1 gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 2 gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 3gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
**

**Comment by saikumarchalla** [comment]
@Neronjust2017 Could you please fill the issue template.Also Please provide the simple standalone code/ colab link to reproduce the issue at our end.Thanks

**Comment by Neronjust2017** [comment]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):  yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Red Hat Enterprise Linux Server release 7.9 (Maipo)
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): gpu 1.14.0
- Python version: 3.7.10
- CUDA/cuDNN version: cuda 10.2 cudnn 8.2
- GPU model and memory: Tesla P100-PCIE 16GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I am working on deploy some yolo v3 models (keras) on the GPU of our server. However, I found that, when I use multithreading programming in python to deploy multiple models and let them predict in parallel, the GPU memory usage does not increase, the GPU memory usage is almost as much as one single model, which makes me very confused. 

load models one by one and measure the gpu memory: 

model 0
gpu usage：total 16280.88 MB， used 859.56 MB， unused 15421.31 MB
model 1
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB
model 2
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB
model 3
gpu usage：total 16280.88 MB， used 1371.56 MB， unused 14909.31 MB


infer one by one and measure the gpu memory: 

gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB
gpu usage：total 16280.88 MB， used 6649.56 MB， unused 9631.31 MB


infer with multithreading and measure the gpu memory: 

thread 0  gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 1  gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 2  gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB
thread 3 gpu usage：total 16280.88 MB， used 6651.56 MB， unused 9629.31 MB


**Standalone code to reproduce the issue**

load and predict one by one:

`import os
import time
import numpy as np
from keras.models import Model, load_model
import tensorflow as tf
from keras.backend.tensorflow_backend import set_session
import pynvml
import cv2

os.environ['CUDA_VISIBLE_DEVICES'] = '0'
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)
set_session(tf.Session(config=config))

pynvml.nvmlInit()
handler = pynvml.nvmlDeviceGetHandleByIndex(0)

model_files= [
    '/data/yolo_1',
    '/data/yolo_2',
    '/data/yolo_3',
    '/data/yolo_4'
]

models = []
for i in range(len(model_files)):
    model = load_model(model_files[i])
    models.append(model)
    print(i)
    meminfo = pynvml.nvmlDeviceGetMemoryInfo(handler)
    gpu_mem_total = round(meminfo.total / 1024 / 1024, 2)
    gpu_mem_used = round(meminfo.used / 1024 / 1024, 2)
    gpu_mem_free = round(meminfo.free / 1024 / 1024, 2)
    print(r'gpu memory：total {} MB， used {} MB， unused {} MB'.format(gpu_mem_total, gpu_mem_used, gpu_mem_free))

image_paths = []

input_path = "/data/testing/"

if os.path.isdir(input_path):
    for inp_file in os.listdir(input_path):
        image_paths += [input_path + inp_file]
else:
    image_paths += [input_path]

image_paths = [inp_file for inp_file in image_paths if (inp_file[-4:] in ['.jpg', '.png', 'JPEG'])]

new_w, new_h = 832, 832

t1 = time.time()

for i ,model in enumerate(models):

    for image_path in image_paths:
        image = cv2.imread(image_path)
        image = cv2.resize(image / 255., (new_w, new_h))
        image = np.expand_dims(image, axis=0)
        boxes = model.predict(image)

    meminfo = pynvml.nvmlDeviceGetMemoryInfo(handler)
    gpu_mem_total = round(meminfo.total / 1024 / 1024, 2)
    gpu_mem_used = round(meminfo.used / 1024 / 1024, 2)
    gpu_mem_free = round(meminfo.free / 1024 / 1024, 2)
    print(r'gpu memory：total {} MB， used {} MB， unused {} MB'.format(gpu_mem_total, gpu_mem_used, gpu_mem_free))

t2 = time.time()
print(t2-t1)
`
predict with multi threading:

`import os
import numpy as np
from keras.models import Model, load_model
import tensorflow as tf
from keras.backend.tensorflow_backend import set_session
import pynvml
import cv2
import threading
import time

os.environ['CUDA_VISIBLE_DEVICES'] = '0'
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)
set_session(tf.Session(config=config))

pynvml.nvmlInit()
handler = pynvml.nvmlDeviceGetHandleByIndex(0)

model_files= [
    '/data/yolo_1',
    '/data/yolo_2',
    '/data/yolo_3',
    '/data/yolo_4'
]

image_paths = []

input_path = "/data/testing/"

if os.path.isdir(input_path):
    for inp_file in os.listdir(input_path):
        image_paths += [input_path + inp_file]
else:
    image_paths += [input_path]

image_paths = [inp_file for inp_file in image_paths if (inp_file[-4:] in ['.jpg', '.png', 'JPEG'])]
print(len(image_paths))

models = []
for i in range(len(model_files)):
    model = load_model(model_files[i])
    model._make_predict_function()
    models.append(model)
    print(i)
    meminfo = pynvml.nvmlDeviceGetMemoryInfo(handler)
    gpu_mem_total = round(meminfo.total / 1024 / 1024, 2)
    gpu_mem_used = round(meminfo.used / 1024 / 1024, 2)
    gpu_mem_free = round(meminfo.free / 1024 / 1024, 2)
    print(r'gpu memory：total {} MB， used {} MB， unused {} MB'.format(gpu_mem_total, gpu_mem_used, gpu_mem_free))


def infer(i):

    new_w, new_h = 832, 832

    for image_path in image_paths:
        image = cv2.imread(image_path)
        image = cv2.resize(image / 255., (new_w, new_h))
        image = np.expand_dims(image, axis=0)
        boxes = models[i].predict(image)

    meminfo = pynvml.nvmlDeviceGetMemoryInfo(handler)
    gpu_mem_total = round(meminfo.total / 1024 / 1024, 2)
    gpu_mem_used = round(meminfo.used / 1024 / 1024, 2)
    gpu_mem_free = round(meminfo.free / 1024 / 1024, 2)
    print(r'{} gpu memory：total {} MB， used {} MB， unused {} MB'.format(i, gpu_mem_total, gpu_mem_used, gpu_mem_free))


threads = []
for i, model_file in enumerate(model_files):
    threads.append(threading.Thread(target=infer, args=(i,)))

for thread in threads:
    thread.start()
`

**Comment by Saduf2019** [comment]
@Neronjust2017 
We see that you are using a old version of tensorflow [1.x] that is not supported now, please upgrade to tensorflow 2.x and let us know if you face any issues.
